// tsparticles-engine/browser/Options/Classes/Particles/Move/Move.js
import { MoveAngle } from "./MoveAngle";
import { MoveAttract } from "./MoveAttract";
import { MoveGravity } from "./MoveGravity";
import { MovePath } from "./Path/MovePath";
import { MoveTrail } from "./MoveTrail";
import { OutModes } from "./OutModes";
import { Spin } from "./Spin";
import { deepExtend } from "../../../../Utils/Utils";
import { setRangeValue } from "../../../../Utils/NumberUtils";
/**
 * [[include:Options/Particles/Move.md]]
 * @category Options
 */
export class Move {
    constructor() {
        this.angle = new MoveAngle();
        this.attract = new MoveAttract();
        this.center = {
            x: 50,
            y: 50,
            mode: "percent" /* SizeMode.percent */,
            radius: 0,
        };
        this.decay = 0;
        this.distance = {};
        this.direction = "none" /* MoveDirection.none */;
        this.drift = 0;
        this.enable = false;
        this.gravity = new MoveGravity();
        this.path = new MovePath();
        this.outModes = new OutModes();
        this.random = false;
        this.size = false;
        this.speed = 2;
        this.spin = new Spin();
        this.straight = false;
        this.trail = new MoveTrail();
        this.vibrate = false;
        this.warp = false;
    }
    /**
     * @deprecated this property is obsolete, please use the new collisions object on particles options
     */
    get bounce() {
        return this.collisions;
    }
    /**
     * @deprecated this property is obsolete, please use the new collisions object on particles options
     * @param value
     */
    set bounce(value) {
        this.collisions = value;
    }
    /**
     * @deprecated this property is obsolete, please use the new collisions object on particles options
     */
    get collisions() {
        return false;
    }
    /**
     * @deprecated this property is obsolete, please use the new collisions object on particles options
     * @param _
     */
    set collisions(_) {
        // deprecated
    }
    /**
     * @deprecated use the new [[path]] property instead
     */
    get noise() {
        return this.path;
    }
    /**
     * @deprecated use the new [[path]] property instead
     */
    set noise(value) {
        this.path = value;
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new outMode
     */
    get outMode() {
        return this.outModes.default;
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new outMode
     * @param value
     */
    set outMode(value) {
        this.outModes.default = value;
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new outMode
     */
    get out_mode() {
        return this.outMode;
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new outMode
     * @param value
     */
    set out_mode(value) {
        this.outMode = value;
    }
    load(data) {
        var _a, _b, _c, _d;
        if (!data) {
            return;
        }
        this.angle.load(typeof data.angle === "number" ? { value: data.angle } : data.angle);
        this.attract.load(data.attract);
        this.center = deepExtend((_a = this.center) !== null && _a !== void 0 ? _a : {}, data.center);
        if (data.decay !== undefined) {
            this.decay = data.decay;
        }
        if (data.direction !== undefined) {
            this.direction = data.direction;
        }
        if (data.distance !== undefined) {
            this.distance =
                typeof data.distance === "number"
                    ? {
                        horizontal: data.distance,
                        vertical: data.distance,
                    }
                    : Object.assign({}, data.distance);
        }
        if (data.drift !== undefined) {
            this.drift = setRangeValue(data.drift);
        }
        if (data.enable !== undefined) {
            this.enable = data.enable;
        }
        this.gravity.load(data.gravity);
        const outModes = (_c = (_b = data.outModes) !== null && _b !== void 0 ? _b : data.outMode) !== null && _c !== void 0 ? _c : data.out_mode;
        if (outModes !== undefined) {
            if (typeof outModes === "object") {
                this.outModes.load(outModes);
            }
            else {
                this.outModes.load({
                    default: outModes,
                });
            }
        }
        this.path.load((_d = data.path) !== null && _d !== void 0 ? _d : data.noise);
        if (data.random !== undefined) {
            this.random = data.random;
        }
        if (data.size !== undefined) {
            this.size = data.size;
        }
        if (data.speed !== undefined) {
            this.speed = setRangeValue(data.speed);
        }
        this.spin.load(data.spin);
        if (data.straight !== undefined) {
            this.straight = data.straight;
        }
        this.trail.load(data.trail);
        if (data.vibrate !== undefined) {
            this.vibrate = data.vibrate;
        }
        if (data.warp !== undefined) {
            this.warp = data.warp;
        }
    }
}
