// tsparticles-engine/browser/Options/Classes/Interactivity/Interactivity.js
import { Events } from "./Events/Events";
import { Modes } from "./Modes/Modes";
/**
 * [[include:Options/Interactivity.md]]
 * @category Options
 */
export class Interactivity {
    constructor(engine, container) {
        this.detectsOn = "window" /* InteractivityDetect.window */;
        this.events = new Events();
        this.modes = new Modes(engine, container);
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new detectsOn
     */
    get detect_on() {
        return this.detectsOn;
    }
    /**
     *
     * @deprecated this property is obsolete, please use the new detectsOn
     * @param value
     */
    set detect_on(value) {
        this.detectsOn = value;
    }
    load(data) {
        var _a;
        if (!data) {
            return;
        }
        const detectsOn = (_a = data.detectsOn) !== null && _a !== void 0 ? _a : data.detect_on;
        if (detectsOn !== undefined) {
            this.detectsOn = detectsOn;
        }
        this.events.load(data.events);
        this.modes.load(data.modes);
    }
}
